dandiapi/api/services/dandiset/__init__.py

```diff
@@ -74,3 +78,39 @@ def delete_dandiset(*, user, dandiset: Dandiset) -> None:
 
         dandiset.versions.all().delete()
         dandiset.delete()
+
+
+def star_dandiset(*, user, dandiset: Dandiset) -> int:
+    """
+    Star a Dandiset for a user.
+
+    Args:
+        user: The user starring the Dandiset.
+        dandiset: The Dandiset to star.
+
+    Returns:
+        The new star count for the Dandiset.
+    """
+    if not user.is_authenticated:
+        raise NotAuthenticatedError
+
+    DandisetStar.objects.get_or_create(user=user, dandiset=dandiset)
+    return dandiset.star_count
+
+
+def unstar_dandiset(*, user, dandiset: Dandiset) -> int:
+    """
+    Unstar a Dandiset for a user.
+
+    Args:
+        user: The user unstarring the Dandiset.
+        dandiset: The Dandiset to unstar.
+
+    Returns:
+        The new star count for the Dandiset.
+    """
+    if not user.is_authenticated:
+        raise NotAuthenticatedError
+
+    DandisetStar.objects.filter(user=user, dandiset=dandiset).delete()
```

Comment:
Similar to the previous suggestion about `star_dandiset`. This would mainly be catching the case where the function gets called on a dandiset that does not have a star from this user (i.e. when `objects_deleted` is `0`)
```suggestion
    objects_deleted = DandisetStar.objects.filter(user=user, dandiset=dandiset).delete()
    if objects_deleted != 1:
        # replace with a custom exception class that propogates a 4xx error up to the view response
        raise Exception(f'{objects_deleted} objects deleted by unstar_dandiset!!')
```

Reply:
Resolution - https://github.com/dandi/dandi-archive/pull/2123#issuecomment-2625262813